### ISSUE TITLE
Skill usage and charge attack toggle button issues for non-Viramate users

### ISSUE THREAD

**Issue by MagCat** [Masuzu/ZooeyBot#32]
Hi, I'm new to lua script and trying to grasp the basics of this. My farming team has Yohda in 3rd position and I have trouble to get his combat behavior to act in certain condition. All of the simple action works fine but anything involves with if and then results in zooeybot crashing. This is what I'm having.

Summon(6)
character_1:UseSkill(4)
character_2:UseSkill(2)
character_4:UseSkill(3)
if character_3.charge_gauge >= 100 then
  character_3:UseSkill(3)
  end
if character_3.charge_gauge >= 100 then
  EnableChargeAttack()
  else
  DisableChargeAttack()
  end

Alternatively, I tried to combine EnableChargeAttack() with first command and it still results in crashing. Like this.
if character_3.charge_gauge >= 100 then
  character_3:UseSkill(3)
  EnableChargeAttack()
  else
  DisableChargeAttack()
  end

**Comment by Masuzu** [comment]
Hi,
What do you mean by "crashing"?
Would you have an error message?

**Comment by MagCat** [comment]
There is no message when it crashes except ".exe stopped working." Everything just crashes together without leaving error.

**Comment by Masuzu** [comment]
It looks like a bug with the charge_gauge  variable then, as I am now using some conditions just fine.
I will investigate this and provide a fix.

**Comment by Masuzu** [comment]
Just a little question please:
`the client just crashes whenever it tries to go back and enable Charge Attack.`
Do you mean that that it pressed on the "back" button?

**Comment by MagCat** [comment]
I think it tried to click on the "back" button but it couldn't, and just crashed. But yeah, the "back" button wasn't pressed.

**Comment by Masuzu** [comment]
Ah I identified the issue.
In the meantime you should use instead:
```
if character_3.charge_gauge >= 100 then
  EnableChargeAttack()
  character_3:UseSkill(3)
end
```

**Comment by MagCat** [comment]
You are probably already working on this issue but I encountered a new issue. I altered your script little by adding new condition, and while it's working fine for 1 or 2 times, the client still consistently crashes after 2-3 runs. 

if battle == 3 and character_3.charge_gauge >= 100 then
  EnableChargeAttack()
  character_3:UseSkill(3)
end

There is another issue I don't know if I should make a new thread, but sometimes the bot gets stuck in message "you're debuffed and can't use skills." This happens randomly.

**Comment by almartn123** [comment]
Since this page is asking for .lua help might as well post here.
Is there a way to delay the bot for a few seconds? Usually after full-chaining there's a ~16s delay before another action. Since the bot immediately refreshes after attack command, the game is still on wait as the bot tries to do more actions missing up my order of skill commands on turns causing it to go ahead than normal.

**Comment by Masuzu** [comment]
@almartn123 After each action (skill or summon usage) following an Ougi, a popup window saying that you can't use the skill or summon yet should appear. As long as it does, the bot should try to reuse the same skill or summon which it could not due to this "action lock out" pop up error message.

**Comment by MagCat** [comment]
Actually almartn123 made a very interesting point, because after I looked at how my script to farm Aurora Trial, I notice a very consistent detail that leads to crashing. Since I made my protag keeps using a skill (4) whenever it's available, whenever the bot tries to switch to Yohda's condition (if charge gauge >= 100) and uses a skill to enable charge attack, the bot just crashes consistently. After I delete the line to make protag not use skill, I haven't noticed crashes so far. Before that it always crashes on 1st or 3rd repeat.

DisableChargeAttack()
Summon(6)
**character_1:UseSkill(4)**
character_2:UseSkill(2)

if battle == 2 and turn == 1 then
character_1:UseSkill(3)
character_1:UseSkill(2)
character_1:UseSkill(1)
end

if battle == 3 and character_3.charge_gauge >= 100 then
  EnableChargeAttack()
  character_3:UseSkill(3)
end

**Comment by Masuzu** [comment]
@MagCat Ah I have the complete picture now. I thought that there was no skill usage before you tried to call `EnableChargeAttack`.
The main issue if you do not use Viramate hotkeys (for now) is that Zooey won't try to press the "Back button" whenever it tries to EnableChargeAttack or DisableChargeAttack, thus leading to the crash.
The next version will patch this issue.
Have EnableChargeAttack() called before any ability is used will work though.

**Comment by MagCat** [comment]
Another crash right now which I can't fathom why it's such a big problem is when the message "you're debuffed and can't use skills" pop up, the bot either locks itself because it can't click on attack button, or it just plain out crashes in middle of the fight. 
almartn123 has a pretty reasonable explanation, but the debuffed message pops up during normal fights as well. Or even at the beginning of the battle **BEFORE** bot could click on any skills.

**Comment by Masuzu** [comment]
That's a pretty interesting case to have the "you're debuffed and can't use skills" pop up" showing up right after refreshing the page o.O
Thank you for reporting this "corner case".
Hopefully the fix is not difficult for this case and I will provide a patched version in a few hours from now.

**Comment by Masuzu** [comment]
The crash following the attempt to click on the Enable/Disable charge attack should be fixed with the version 1.3.0 ([patch notes](https://poker-bot.azurewebsites.net/ZooeyBot/en/Home/PatchNotes). To test it, I used a simple script where I used a skill first before enabling/disabling the charge attack.

As for the "you're debuffed and can't use skills" pop up showing up right after refreshing the page o.O, I would need more info about it. I assume that when it appears, the attack button is visible but Zooey can't do anything because it is overlapped by the error window, is that right? Under this assumption, the version 1.3.0 should also provide a fix for this case.

When you try to use a skill while you can't (e.g. Vira 4th skill without enough charge bar), a similar error message will appear too. In this case Zooey will try to dismiss it up to `MaxNumActionRetries` times before continuing with the rest of the command list scripted by the Lua script. (To be updated later on to retry only once).

**Comment by MagCat** [comment]
After a few test with the 1.3.0 version, I haven't got any crashes and debuff message while getting the bots to do more commands simultaneously now. I still need to test with this version in more scenarios but this REALLY HELPS a lot!

**Comment by Masuzu** [comment]
Thank you for the feedback @MagCat , much appreciated.
I will leave this ticket open for another few days.

**Comment by MagCat** [comment]
Okay I got the accursed debuffed message again. This time I monitored it personally in 2 battles and I found something is interesting. 

Bug (debuffed message). I found out that when a character uses an attack skill that BREAK monster, a battle log says "Break took effect! blah blah" and the debuff message shows up immediately right after. However, this bug proves to be even more troublesome when the message shows up at beginning battle when bot is changing from using one character's skill to another's. 
I don't know if this will be possible but I suggest a reload command when things not respond for 5s or such.

**Comment by MagCat** [comment]
Seriously, I figured out that since I didn't have keyboard option activated in Viramate's option, ZooeyBot didn't know what to do so it locked itself in battle. Now that I know how it works this is much better than not using viratemate. not only it's almost bug free the battles are also much faster in general.
